Completed homework 16 sprint

RURURuina opened the pull request:
Заранее спасибо за проверку <3 
Прошу писать подробные комментарии, пожалуйста)

RURURuina committed:
Completed homework 16 sprint
Заранее спасибо за проверку <3

IliaAnK commented on app/src/main/java/com/practicum/playlistmaker/search/ui/SearchActivity.kt:
```diff
@@ -71,6 +68,9 @@ class SearchActivity : AppCompatActivity() {
+        val factory = SearchViewModelFactory(application)
+        viewModel = ViewModelProvider(this, factory).get(SearchViewModel::class.java)
+
         val sharedPreferences = getSharedPreferences("SearchHistory", MODE_PRIVATE)
```
Обязательно:
Работа с SharedPreferences должна происходить исключительно в `data` слое
Нужно перенести в репозиторий и выстроить взаимодействие с полем согласно клин архитектуре

IliaAnK commented on app/src/main/java/com/practicum/playlistmaker/search/domain/models/Track.kt:
```diff
@@ -1,4 +1,4 @@
-package com.practicum.playlistmaker.domain.models
+package com.practicum.playlistmaker.search.domain.models
 
 import android.os.Parcel
```
Обязательно:
Классы в пакете `domain` не должны использовать классы из Android SDK.
Здесь можно заменить на `Serializable` или выделить презентационную модель и добавить маппинг между `presentation` и `domain` слоями.

На мой взгляд, с точки зрения чистой архитектуры, второй вариант предпочтительнее

IliaAnK commented on app/src/main/java/com/practicum/playlistmaker/search/ui/viewmodel/SearchViewModel.kt:
```diff
@@ -0,0 +1,40 @@
+        }
+    }
+
+    fun getSearchHistory(): MutableList<Track?> {
```
Обязательно:
`Activity` должна получать все необходимые для отображения данные, подписываясь на `LiveData`-у.
`ViewModel` в свою очередь обновляет `LiveData`-у и не возвращает данные напрямую.

IliaAnK commented on app/src/main/java/com/practicum/playlistmaker/player/ui/view_model/MediaPlayerViewModel.kt:
```diff
@@ -0,0 +1,47 @@
+        _currentPosition.value = interactor.getCurrentPosition()
+    }
+
+    fun isPlaying(): Boolean {
```
Обязательно:
`Activity` должна получать все необходимые для отображения данные, подписываясь на `LiveData`-у.
`ViewModel` в свою очередь обновляет `LiveData`-у и не возвращает данные напрямую.

IliaAnK commented on app/src/main/java/com/practicum/playlistmaker/player/ui/view_model/MediaPlayerViewModel.kt:
```diff
@@ -0,0 +1,47 @@
+
+class MediaPlayerViewModel(private val interactor: MediaPlayerInteractor) : ViewModel() {
+
+    private val _playerState = MutableLiveData<PlayerState>().apply { value = PlayerState.DEFAULT }
```
Обязательно:
Не должно быть отдельной `LiveData` для каждой `View`. 
Нужно объединить значения из полей `_playerState` и `_currentPosition` в один data и/или sealed класс, тем самым сформировать единый стейт экрана.

RURURuina committed:
Completed homework 16 sprint
Суммарно 15 часов внесения правок в код))))))
Было исправлено:
1. Баг при котором таймер не обнулялся.
2. Цвет прогрес бара
3. Вся работа с SharedPreferences происходит в data слое. В ui слое её больше нет.
4. В классе Track метод андроид sdk Parcel заменен на Serializable
5. Activity получает все данные подписываясь на LiveData
6. В AudioPlayerActivity пришлось много править код, чтобы метод isPlaying() использовался только в viewModel, пытался сделать без этого метода, что-то вроде if(_playerState.value == PlayerState.PLAYING), но при использовании такого способа, переставал работать таймер и кнопка паузы
(просидел с этим багом больше всего времени))))
7. _playerState и _playerCurrentPosition объединены в _playerState по средством sealed interface PlayerState
Заранее спасибо за проверку <3
